scylla/src/transport/connection.rs

```diff
@@ -957,22 +962,36 @@ impl Connection {
     }
 
     // This task receives notifications from `OrphanhoodNotifier`s and tries to
-    // mark streams as orphaned
+    // mark streams as orphaned. After an orphan threshold is reached, it
+    // returns an error causing the connection to shutdown.
     async fn orphaner(
         handler_map: &StdMutex<ResponseHandlerMap>,
         mut orphan_receiver: mpsc::UnboundedReceiver<GenerationalStreamId>,
     ) -> Result<(), QueryError> {
-        while let Some(gsi) = orphan_receiver.recv().await {
-            trace!(
-                "Trying to orphan stream_id = {} with generation = {}",
-                gsi.stream_id,
-                gsi.generation
-            );
-            // We are guaranteed here that handler_map will not be locked
-            // by anybody else, so we can do try_lock().unwrap()
-            let mut lock = handler_map.try_lock().unwrap();
-            lock.orphan(gsi);
+        let mut interval = tokio::time::interval(OLD_AGE_ORPHAN_THRESHOLD);
+        loop {
+            tokio::select! {
+                _ = interval.tick() => {
+                    // We are guaranteed here that handler_map will not be locked
+                    // by anybody else, so we can do try_lock().unwrap()
+                    let lock = handler_map.try_lock().unwrap();
+                    let old_orphan_count = lock.old_orphan_count();
+                    if old_orphan_count > OLD_ORPHAN_COUNT_THRESHOLD {
+                        warn!(
+                            "Too many old orphaned stream ids: {}",
+                            old_orphan_count,
+                        );
+                        return Err(QueryError::TooManyOrphanedStreamIds(old_orphan_count as u16))
+                    }
+                }
+                Some(gsi) = orphan_receiver.recv() => {
+                    let mut lock = handler_map.try_lock().unwrap(); // Same as above
```

Comment:
nitpick: lock is a terrible name for this variable, because it's not a lock anymore, it's a map; please change it to something more descriptive, preferably ending with a `_map`

Reply:
`lock` is of type [`MutexGuard`](https://doc.rust-lang.org/std/sync/struct.MutexGuard.html), which is a kind of a lock. Maybe `guard`, `map_guard` or something similar will be more clear, then?

Reply:
@piodul technically it's a lock that gets dereferenced automatically, but calling this variable a "lock" would be equivalent to calling a C++ variable, which is a pointer to something, a "pointer".  `guard` sounds similarly non-descriptive to me, `map_guard` is better, `handler_map_guard` is better yet.